image/mae/main_pretrain.py

```diff
@@ -0,0 +1,236 @@
+    train_dataset = _load_imgfolder(datapath / "train", transform_train)
+    test_dataset = _load_imgfolder(datapath / "val", transform_val)
+    if VERBOSE:
+        LOGGER.info(f"Train dataset:\n{train_dataset}")
+        LOGGER.info(f"Test dataset:\n{test_dataset}")
+
+    train_dataloader = get_dataloader(
+        dataset=train_dataset,
+        shuffle=True,
+        batch_size=gpc.config.BATCH_SIZE,
+        num_workers=1,
+        pin_memory=True,
+        drop_last=True,
+    )
+
+    test_dataloader = get_dataloader(
+        dataset=train_dataset,
+        shuffle=True,
+        batch_size=gpc.config.BATCH_SIZE,
+        num_workers=1,
+        pin_memory=True,
+        drop_last=False,
+    )
+
+    return train_dataloader, test_dataloader
+
+
+def init_global_states(config: Config):
+    global VERBOSE
+    VERBOSE = config.VERBOSE
+
+
+def init_engine(config: Config):
+    _model = model(config.NORM_PIX_LOSS)
+    _optimizer = optimizer(_model, config.LEARNING_RATE, config.WEIGHT_DECAY)
+    _criterion = criterion()
+    train_dataloader, test_dataloader = pretrain_dataloaders(
+        config.DATAPATH, config.TRANSFORM_TRAIN, config.TRANSFORM_VAL
+    )
+    engine, train_dataloader, test_dataloader, _ = colossalai.initialize(
+        _model,
+        _optimizer,
+        _criterion,
+        train_dataloader,
+        test_dataloader,
+    )
+    return engine, train_dataloader, test_dataloader
+
+
+def resume_model(engine, loss_scaler, config):
+    args = load_model_args(
+        resume=config.RESUME_ADDRESS, start_epoch=config.RESUME_START_EPOCH
+    )
+
+    misc.load_model(
+        args=args,
+        model_without_ddp=engine.model,
+        optimizer=engine.optimizer,
+        loss_scaler=loss_scaler,
+    )
+    if VERBOSE:
+        LOGGER.info(
+            f"Resume model from {config.RESUME_ADDRESS}, start at epoch {config.RESUME_START_EPOCH}"
+        )
+
+    return args.start_epoch
+
+
+def adjust_learning_rate(engine, data_iter_step, train_dataloader, epoch, config):
+    args = lr_sched_args(
+        lr=config.LEARNING_RATE,
+        min_lr=config.MINIMUM_LEARNING_RATE,
+        epochs=config.NUM_EPOCHS,
+        warmup_epochs=config.WARMUP_EPOCHS,
+    )
+    lr_sched.adjust_learning_rate(
+        engine.optimizer,
+        data_iter_step / len(train_dataloader) + epoch,
+        args,
+    )
+
+
+def exit_if_infinite_loss(l):
+    if not math.isfinite(l):
+        print("Loss is {}, stopping training".format(l))
+        sys.exit(1)
+
+
+def scale_loss(engine, loss, loss_scaler, data_iter_step, config):
+    loss /= config.ACCUM_ITER
+    loss_scaler(
+        loss,
+        engine.optimizer,
+        parameters=engine.model.parameters(),
+        update_grad=(data_iter_step + 1) % config.ACCUM_ITER == 0,
+    )
+
+
+def save_model(model, output_dir, epoch, config, optimizer, loss_scaler):
+    checkpoint_path = output_dir / (f"checkpoint-{epoch}.pth")
+    torch.save(
```

Comment:
the model saving just works for DP....

Reply:
Well, since I've implemented TP, I am still not sure the way of saving model under TP... Are there any examples about it?

Reply:
Moved to `colossalai.utils.checkpointing`.